File: prototype/public/assets/signal.js
/* Signalling transport.
 *
 * Server-Sent Events downstream, plain POST upstream. No WebSocket, no library.
 * This carries offer / answer / ICE only — never media.
 *
 * Two secrets are in play:
 *   - the pairing proof, sha256(uid + ':' + PIN), which authorises joining a room
 *   - a per-connection bearer token, handed back on our own event stream, which
 *     must accompany every POST. It binds each POST to a live connection, so a
 *     third party can neither impersonate a peer nor forge requests from
 *     another origin. */

(function (global) {
  'use strict';

  function randomId() {
    var b = new Uint8Array(16);
    global.crypto.getRandomValues(b);
    var s = '';
    for (var i = 0; i < b.length; i++) s += b[i].toString(16).padStart(2, '0');
    return s;
  }

  function Signal(uid, role, auth) {
    this.uid = String(uid).toUpperCase();
    this.role = role;
    this.auth = auth || '';
    this.id = randomId();
    this.token = null;
    this.handlers = {};
    this.es = null;
    this.peers = [];
    this.failures = 0;
    this.closed = false;
  }

  Signal.prototype.on = function (ev, fn) {
    (this.handlers[ev] = this.handlers[ev] || []).push(fn);
    return this;
  };

  Signal.prototype.emit = function (ev, data) {
    (this.handlers[ev] || []).forEach(function (fn) {
      try { fn(data); } catch (err) { console.error('[signal]', ev, err); }
    });
  };

  /* Side-effect-free pairing probe. Resolves { ok: true } or { ok: false,
   * status, error } so the caller can show a real message — EventSource itself
   * never exposes the HTTP status. */
  Signal.check = function (uid, auth) {
    var url = '/api/check?uid=' + encodeURIComponent(String(uid).toUpperCase()) +
              '&auth=' + encodeURIComponent(auth || '');
    return fetch(url, { credentials: 'omit' }).then(function (r) {
      return r.json().catch(function () { return {}; }).then(function (body) {
        return r.ok ? { ok: true } : { ok: false, status: r.status, error: body.error || ('HTTP ' + r.status) };
      });
    }).catch(function (err) {
      return { ok: false, status: 0, error: 'cannot reach the server (' + err.message + ')' };
    });
  };

  Signal.prototype.connect = function () {
    var self = this;
    var url = '/api/events?uid=' + encodeURIComponent(this.uid) +
              '&role=' + encodeURIComponent(this.role) +
              '&id=' + encodeURIComponent(this.id) +
              '&auth=' + encodeURIComponent(this.auth);

    var es = new EventSource(url);
    this.es = es;

    ['ready', 'hello', 'bye', 'roster', 'signal'].forEach(function (name) {
      es.addEventListener(name, function (e) {
        var data;
        try { data = JSON.parse(e.data); } catch (err) { return; }
        if (name === 'ready') { self.token = data.token; self.failures = 0; }
        if (name === 'roster') self.peers = data.peers || [];
        self.emit(name, data);
      });
    });

    es.onopen = function () { self.failures = 0; self.emit('open'); };

    es.onerror = function () {
      if (self.closed) return;
      self.failures += 1;
      /* EventSource retries forever by default, which against a 403 is an
       * accidental PIN-guessing loop. Give up after a few. */
      if (self.failures >= 4) {
        self.close();
        self.emit('failed', { attempts: self.failures });
      } else {
        self.emit('down', { attempts: self.failures });
      }
    };

    return this;
  };

  Signal.prototype.send = function (type, data, to) {
    if (!this.token) return Promise.resolve({ error: 'not authorised yet' });
    var body = JSON.stringify({
      uid: this.uid, from: this.id, token: this.token,
      to: to || null, type: type, data: data,
    });
    return fetch('/api/signal', {
      method: 'POST',
      credentials: 'omit',
      headers: { 'content-type': 'application/json' },
      body: body,
    }).then(function (r) { return r.json().catch(function () { return {}; }); })
      .catch(function (err) { return { error: String(err) }; });
  };

  Signal.prototype.close = function () {
    this.closed = true;
    if (this.es) { this.es.close(); this.es = null; }
  };

  global.Signal = Signal;
  global.randomId = randomId;
})(window);

File: android/app/src/main/java/ai/less/p2pcam/net/Signal.kt
package ai.less.p2pcam.net

import android.os.Handler
import android.os.Looper
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.sse.EventSource
import okhttp3.sse.EventSourceListener
import okhttp3.sse.EventSources
import org.json.JSONObject
import java.util.concurrent.TimeUnit
import javax.net.ssl.HostnameVerifier
import javax.net.ssl.SSLContext
import javax.net.ssl.X509TrustManager

/**
 * Signalling transport: Server-Sent Events down, POST up. Carries offer,
 * answer and ICE only — never media.
 *
 * Two secrets are in play, mirroring the browser client:
 *   - the pairing proof, sha256(uid + ':' + PIN), which authorises joining
 *   - a per-connection bearer token, delivered only on this peer's own event
 *     stream, which must accompany every POST
 */
class Signal(
    private val baseUrl: String,
    private val uid: String,
    private val role: String,
    private val auth: String,
    allowSelfSigned: Boolean,
) {

    interface Listener {
        fun onReady(token: String) {}
        fun onHello(from: String, peerRole: String) {}
        fun onBye(from: String, peerRole: String) {}
        fun onRoster(peers: List<Peer>) {}
        fun onSignal(from: String, type: String, data: JSONObject) {}
        fun onDown(attempts: Int) {}
        fun onFailed(reason: String) {}
    }

    data class Peer(val id: String, val role: String)

    val clientId: String = Crypto.randomId()
    var token: String? = null
        private set

    private val main = Handler(Looper.getMainLooper())
    private val client = buildClient(allowSelfSigned)
    private var source: EventSource? = null
    private var failures = 0
    private var closed = false
    private var listener: Listener? = null

    fun listener(l: Listener) = apply { listener = l }

    // ---------------------------------------------------------------- probe --

    data class CheckResult(val ok: Boolean, val status: Int, val error: String?)

    /**
     * Side-effect-free pairing probe. SSE surfaces no status code, so we ask
     * first and can report "wrong PIN" instead of retrying against a 403.
     * Blocking — call from a background thread.
     */
    fun check(): CheckResult {
        val url = "$baseUrl/api/check?uid=${enc(uid)}&auth=${enc(auth)}"
        return try {
            client.newCall(Request.Builder().url(url).get().build()).execute().use { res ->
                val body = res.body?.string().orEmpty()
                if (res.isSuccessful) {
                    CheckResult(true, res.code, null)
                } else {
                    val msg = runCatching { JSONObject(body).optString("error") }.getOrNull()
                    CheckResult(false, res.code, msg?.takeIf { it.isNotBlank() } ?: "HTTP ${res.code}")
                }
            }
        } catch (e: Exception) {
            CheckResult(false, 0, "cannot reach the server (${e.message})")
        }
    }

    // --------------------------------------------------------------- stream --

    fun connect() {
        val url = "$baseUrl/api/events?uid=${enc(uid)}&role=${enc(role)}" +
                "&id=${enc(clientId)}&auth=${enc(auth)}"
        val req = Request.Builder().url(url).header("Accept", "text/event-stream").get().build()

        source = EventSources.createFactory(client).newEventSource(req, object : EventSourceListener() {
            override fun onOpen(eventSource: EventSource, response: okhttp3.Response) {
                failures = 0
            }

            override fun onEvent(eventSource: EventSource, id: String?, type: String?, data: String) {
                val json = runCatching { JSONObject(data) }.getOrNull() ?: return
                main.post { dispatch(type.orEmpty(), json) }
            }

            override fun onFailure(eventSource: EventSource, t: Throwable?, response: okhttp3.Response?) {
                if (closed) return
                failures += 1
                val code = response?.code ?: 0
                // A 4xx will never succeed on retry; do not sit in a loop that
                // is really a PIN-guessing machine.
                val fatal = code in 400..499 || failures >= 4
                main.post {
                    if (fatal) {
                        close()
                        listener?.onFailed(
                            when (code) {
                                403 -> "wrong PIN"
                                404 -> "no camera online with that UID"
                                409 -> "that UID is already claimed"
                                429 -> "too many attempts, wait a minute"
                                else -> t?.message ?: "signalling connection failed"
                            }
                        )
                    } else {
                        listener?.onDown(failures)
                    }
                }
            }
        })
    }

    private fun dispatch(type: String, json: JSONObject) {
        when (type) {
            "ready" -> {
                token = json.optString("token").takeIf { it.isNotBlank() }
                failures = 0
                token?.let { listener?.onReady(it) }
            }
            "hello" -> listener?.onHello(json.optString("from"), json.optString("role"))
            "bye" -> listener?.onBye(json.optString("from"), json.optString("role"))
            "roster" -> {
                val arr = json.optJSONArray("peers") ?: return
                val list = ArrayList<Peer>(arr.length())
                for (i in 0 until arr.length()) {
                    val p = arr.optJSONObject(i) ?: continue
                    list.add(Peer(p.optString("id"), p.optString("role")))
                }
                listener?.onRoster(list)
            }
            "signal" -> listener?.onSignal(
                json.optString("from"),
                json.optString("type"),
                json.optJSONObject("data") ?: JSONObject()
            )
        }
    }

    // ----------------------------------------------------------------- send --

    /** Fire-and-forget; the relay's acknowledgement carries nothing we need. */
    fun send(type: String, data: JSONObject, to: String? = null) {
        val tok = token ?: return
        val body = JSONObject()
            .put("uid", uid)
            .put("from", clientId)
            .put("token", tok)
            .put("to", to ?: JSONObject.NULL)
            .put("type", type)
            .put("data", data)
            .toString()

        val req = Request.Builder()
            .url("$baseUrl/api/signal")
            .post(body.toRequestBody(JSON))
            .build()

        client.newCall(req).enqueue(object : okhttp3.Callback {
            override fun onFailure(call: okhttp3.Call, e: java.io.IOException) = Unit
            override fun onResponse(call: okhttp3.Call, response: okhttp3.Response) { response.close() }
        })
    }

    fun close() {
        closed = true
        source?.cancel()
        source = null
    }

    // -------------------------------------------------------------- plumbing --

    private fun enc(s: String): String = java.net.URLEncoder.encode(s, "UTF-8")

    companion object {
        private val JSON = "application/json; charset=utf-8".toMediaType()

        /**
         * Prototype-only: when the operator opts in, accept the demo server's
         * self-signed certificate. This disables certificate validation for
         * this client and must never ship. Plain HTTP on the LAN is the
         * preferred path and needs none of this.
         */
        private fun buildClient(allowSelfSigned: Boolean): OkHttpClient {
            val b = OkHttpClient.Builder()
                .readTimeout(0, TimeUnit.MILLISECONDS)     // SSE stays open
                .connectTimeout(10, TimeUnit.SECONDS)
                .writeTimeout(15, TimeUnit.SECONDS)
                .retryOnConnectionFailure(true)
                .pingInterval(20, TimeUnit.SECONDS)

            if (allowSelfSigned) {
                val trustAll = object : X509TrustManager {
                    override fun checkClientTrusted(c: Array<out java.security.cert.X509Certificate>?, a: String?) = Unit
                    override fun checkServerTrusted(c: Array<out java.security.cert.X509Certificate>?, a: String?) = Unit
                    override fun getAcceptedIssuers(): Array<java.security.cert.X509Certificate> = arrayOf()
                }
                val ctx = SSLContext.getInstance("TLS")
                ctx.init(null, arrayOf<javax.net.ssl.TrustManager>(trustAll), java.security.SecureRandom())
                b.sslSocketFactory(ctx.socketFactory, trustAll)
                b.hostnameVerifier(HostnameVerifier { _, _ -> true })
            }
            return b.build()
        }
    }
}
